text/0000-stable-rustdoc-urls.md

```diff
@@ -0,0 +1,232 @@
+
+However, Rustdoc does _not_ currently guarantee that making a semver-compatible change to your code will preserve the same URL.
+This means that, for instance, making a type an `enum` instead of a `struct` will change the URL,
+even if your change is in every other way semver-compatible. After this RFC, Rustdoc will guarantee that the URL would stay the same.
+
+The primary motivation for this feature is to allow linking to a semantic version
+of the docs, rather than an exact version. This has several applications:
+
+- [docs.rs] could link to `/package/0.2/path` instead of `/package/0.2.5/path`, making the documentation users see more up-to-date ([rust-lang/docs.rs#1055])
+- blogs could link to exact URLs without fear of the URL breaking ([rust-lang/rust#55160 (comment)][55160-blog])
+- URLs in the standard library documentation would change less often ([rust-lang/rust#55160][55160])
+
+Note that this is a different, but related, use case than [intra-doc links].
+Intra-doc links allow linking consistently in the presence of re-exports for _relative_ links.
+This is intended to be used for _absolute_ links. Additionally, this would allow linking consistently
+outside of Rust code.
+
+[Rustdoc]: https://doc.rust-lang.org/rustdoc/
+[item]: https://doc.rust-lang.org/reference/items.html
+[docs.rs]: https://docs.rs/
+[could link]: https://github.com/rust-lang/docs.rs/issues/1055
+[55160]: https://github.com/rust-lang/rust/issues/55160
+[55160-blog]: https://github.com/rust-lang/rust/issues/55160#issuecomment-680751534
+[intra-doc links]: https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md
+[rust-lang/docs.rs#1055]: https://github.com/rust-lang/docs.rs/issues/1055
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rustdoc will make the following changes to URL structure:
+
+- Item pages will be dependent only on the namespace, not the type of the item.
+
+	Consider the struct `std::process::Command`.
+	Currently, the URL for it looks like `std/process/struct.Command.html`.
+	This RFC proposes to change the URL to `std/process/type.Command.html`.
+	Pages named `kind.name.html` would still be generated (to avoid breaking existing links),
+	but would immediately redirect to the new URL.
+
+- Re-exports will generate a page pointing to the canonical version of the documentation.
+
+	Consider the following Rust code:
+	
+	```rust
+	pub struct Foo;
+	```
+	
+	Rustdoc currently generates a page for this at `struct.Foo.html`.
+	Now, consider what happens when you move the struct to a different module and re-export it
+	(which is a semver-compatible change):
+
+	```rust
+	pub mod foo { pub struct Foo; }
+	pub use foo::Foo;
+	```
+
+	This generates a page at `foo/struct.Foo.html`, but _not_ at `struct.Foo.html`.
+	After this change, rustdoc will generate a page at the top level which redirects
+	to the version nested in the module.
```

Comment:
What if `foo::Foo` is private?

Reply:
Rustdoc would inline the page as normal; there'd be no redirect and the page would only be available at `struct.Foo.html`. I'll add that to the RFC.

Reply:
What if it's re-used more than once, like here:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=823a342228ecd36da6d48e1d1acacc9f

```rust
pub mod a {
    mod b {
        pub struct Foo;
    }
    
    pub mod c {
        pub use super::b::Foo;
    }
    
    pub mod d {
        pub use super::b::Foo;
    }
}

mod f1 {
    use super::a::c::Foo;
    pub fn foo1(f: Foo) { }
}

mod f2 {
    use super::f1::foo1;
    use super::a::d::Foo;
    
    fn foo2(f: Foo) {
        foo1(f)
    }
}
```

I don't know if there's a way to discern that right now with rust-doc, but the compiler knows that `a::c::Foo` and `a::d::Foo` are the same struct, it might be useful to make this information available.

Reply:
That's a separate feature request and not related to this RFC. Both now and after the RFC, the struct will be inlined in both locations - you can tell they're the same struct by looking at `[src]`, which points to the same place in both pages.